src/Workspaces/Core/Portable/Workspace/Solution/ProjectDependencyGraph_RemoveProject.cs

```diff
@@ -0,0 +1,167 @@
+
+            // The direct reverse references map is updated by removing the key for the project getting removed, and
+            // also updating any direct references to the removed project.
+            var reverseReferencesMap = ComputeNewReverseReferencesMapForRemovedProject(
+                existingForwardReferencesMap: _referencesMap,
+                existingReverseReferencesMap: _lazyReverseReferencesMap,
+                projectId);
+            var transitiveReferencesMap = ComputeNewTransitiveReferencesMapForRemovedProject(_transitiveReferencesMap, projectId);
+            var reverseTransitiveReferencesMap = ComputeNewReverseTransitiveReferencesMapForRemovedProject(_reverseTransitiveReferencesMap, projectId);
+            return new ProjectDependencyGraph(
+                projectIds,
+                referencesMap,
+                reverseReferencesMap,
+                transitiveReferencesMap,
+                reverseTransitiveReferencesMap,
+                topologicallySortedProjects: default,
+                dependencySets: default);
+        }
+
+        /// <summary>
+        /// Computes a new <see cref="_referencesMap"/> for the removal of a project.
+        /// </summary>
+        /// <param name="referencesMap">The <see cref="_referencesMap"/> prior to the removal.</param>
+        /// <param name="existingReverseReferencesMap">The <see cref="_lazyReverseReferencesMap"/> prior to the removal.
+        /// This map serves as a hint to the removal process; i.e. it is assumed correct if it contains data, but may be
+        /// omitted without impacting correctness.</param>
+        /// <param name="projectId">The ID of the project which is being removed.</param>
+        /// <returns>The <see cref="_referencesMap"/> for the project dependency graph once the project is removed.</returns>
+        private ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> ComputeNewReferencesMapForRemovedProject(
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>> referencesMap,
+            ImmutableDictionary<ProjectId, ImmutableHashSet<ProjectId>>? existingReverseReferencesMap,
+            ProjectId projectId)
+        {
+            var builder = referencesMap.ToBuilder();
+
+            if (existingReverseReferencesMap is object && existingReverseReferencesMap.TryGetValue(projectId, out var referencingProjects))
+            {
+                // We know all the projects directly referencing 'projectId', so remove 'projectId' from the set of
+                // references in each of those cases directly.
+                foreach (var id in referencingProjects)
+                {
+                    builder[id] = builder[id].Remove(projectId);
+                }
+            }
+            else
+            {
+                // We don't know which projects reference 'projectId', so iterate over all known projects and remove
+                // 'projectId' from the set of references if it exists.
```

Comment:
this almost makes it seem like a pity to not have the maps since they do look like they can speed things up.

Reply:
It's no longer showing up on performance traces. Hoping to get a benchmark in IdeBenchmarks to show the behavior at scale and allow further tuning.